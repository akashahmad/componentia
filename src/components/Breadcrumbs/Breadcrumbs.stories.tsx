import * as React from "react";

import type { Meta, StoryObj } from "@storybook/react";

import { Breadcrumbs, BreadcrumbsItem } from "../Breadcrumbs/Breadcrumbs";

import { HomeIcon } from "../icons/HomeIcon";

const meta: Meta<typeof Breadcrumbs> = {
    component: Breadcrumbs,
    tags: ["autodocs"],
};

export default meta;
type Story = StoryObj<typeof Breadcrumbs>;

const items: BreadcrumbsItem[] = [
    {
        path: "#",
        text: "Home",
    },
    {
        path: "##",
        text: "Projects",
    },
    {
        path: "###",
        text: "Marketing",
    },
];

const itemsWithHomeIcon: BreadcrumbsItem[] = [
    {
        path: "#",
        text: "Home",
        icon: props => <HomeIcon width={14} {...props} />,
    },
    {
        path: "##",
        text: "Projects",
    },
    {
        path: "###",
        text: "Marketing",
    },
];

const BreadcrumbsRenderer = ({ path, text }: BreadcrumbsItem) => {
    return <a href={path}>{text}</a>;
};

export const Default: Story = {
    args: {
        currentItemRenderer: BreadcrumbsRenderer,
        items,
    },
    argTypes: {
        currentItemRenderer: {
            table: {
                type: { summary: "React.FC<BreadcrumbsItem>" },
                defaultValue: { summary: "undefined" },
            },
            control: { type: false },
        },
        items: {
            table: {
                type: { summary: "BreadcrumbsItem[]" },
                defaultValue: { summary: "undefined" },
            },
            control: { type: false },
        },
        className: {
            table: {
                type: { summary: "string" },
                defaultValue: { summary: '""' },
            },
            control: { type: "text" },
        },
        linksTextColor: {
            table: {
                type: { summary: 'hex code, e.g "#FFFFFF"' },
            },
        },
        currentLinkTextColor: {
            table: {
                type: { summary: 'hex code, e.g "#A91320"' },
            },
        },
    },
    render: props => <Breadcrumbs {...props} />,
};

export const WithHomeIcon: Story = {
    args: {
        currentItemRenderer: BreadcrumbsRenderer,
        items: itemsWithHomeIcon,
    },
    argTypes: {},
    render: props => <Breadcrumbs {...props} />,
};

export const WithPrimaryBackground: Story = {
    args: {
        currentItemRenderer: BreadcrumbsRenderer,
        items: itemsWithHomeIcon,
        className: "bg-primary",
        linksTextColor: "white",
        currentLinkTextColor: "white",
    },
    argTypes: {},
    render: props => <Breadcrumbs {...props} />,
};

export const WithTertiaryBackground: Story = {
    args: {
        currentItemRenderer: BreadcrumbsRenderer,
        items: itemsWithHomeIcon,
        className: "bg-black",
        linksTextColor: "white",
        currentLinkTextColor: "white",
    },
    argTypes: {},
    render: props => <Breadcrumbs {...props} />,
};
