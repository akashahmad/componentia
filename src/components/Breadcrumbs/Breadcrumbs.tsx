import * as React from "react";

import cx from "classnames";

import { ChevronRightIcon } from "../icons/ChevronRightIcon";
import { colors } from "../../utils/colors";

export function Breadcrumbs({
    items,
    currentItemRenderer: Renderer,
    linksTextColor = colors.primaryText,
    currentLinkTextColor = colors.primary,
    className = "",
    ...other
}: BreadcrumbsProps) {
    const _getCustomCurrentLinkColor = (index: number) => {
        return items.length !== index + 1
            ? linksTextColor
            : currentLinkTextColor && items.length !== index
            ? currentLinkTextColor
            : colors.primary;
    };

    return (
        <div className={cx("mr flex items-center leading-6 font-medium text-base", className)} {...other}>
            {items.map(({ path, icon: Icon, text }, index) => (
                <div
                    key={path + index}
                    style={{ color: _getCustomCurrentLinkColor(index) }}
                    className={cx("first:pl-[30px] flex items-center mr-3 py-[5px]")}
                >
                    {Icon && (
                        <span className="mr-2">
                            <Icon fill={linksTextColor} />
                        </span>
                    )}
                    <div className={cx("mr-3")}>
                        <Renderer path={path} text={text} icon={Icon} />
                    </div>
                    {items.length !== index + 1 && <ChevronRightIcon fill={linksTextColor} />}
                </div>
            ))}
        </div>
    );
}

export type BreadcrumbsItem = {
    path: string;
    text: string;
    icon?: React.FC<{ fill?: string }>;
};
interface BreadcrumbsProps {
    items: BreadcrumbsItem[];
    currentItemRenderer: React.FC<BreadcrumbsItem>;
    className?: string;
    linksTextColor?: string;
    currentLinkTextColor?: string;
}
