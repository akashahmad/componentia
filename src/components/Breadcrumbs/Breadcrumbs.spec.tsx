import * as React from "react";

import { render } from "@testing-library/react";
import { Breadcrumbs } from "./Breadcrumbs";

test("Breadcrumbs component is rendering properly", () => {
    const { container } = render(<Breadcrumbs items={[]} currentItemRenderer={() => <p></p>} />);

    expect(container).toMatchSnapshot();
});
