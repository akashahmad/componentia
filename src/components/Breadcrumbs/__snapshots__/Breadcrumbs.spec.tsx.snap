// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Breadcrumbs component is rendering properly 1`] = `
<div>
  <div
    class="mr flex items-center leading-6 font-medium text-base"
  />
</div>
`;
